### ISSUE TITLE
Duplicate entries for same product id

### ISSUE THREAD

**Issue by klickreflex** [jackabox/statamic-shopify#95]
After letting my client add products in Shopify I started noticing duplicate item in the Statamic product collection with the same `product_id`.

 Differences I noticed where:
- published status in one case
- image URLs with/without timestamp suffix (like in #86)
- the markdown files were suffixed with a `1` (screenshot 1)
- There are variants in Statamic that don't exist in Shopify (screenshot 2)

I'm using webhooks to update Statamic.

![screenshot 1](https://user-images.githubusercontent.com/435825/139199012-b0dc81b9-7af2-48bb-b3e0-e6108c6d2725.png)

![screenshot 2](https://user-images.githubusercontent.com/435825/139199918-6cc2097a-aeda-45a8-93f3-6392d81e84f7.png)

**Comment by klickreflex** [comment]
This is getting out of hand with more products being added :(
![CleanShot 2021-11-05 at 10 30 58@2x](https://user-images.githubusercontent.com/435825/140489436-841d8e61-d66d-4d0a-a150-f10ac70cc527.png)

**Comment by jackabox** [comment]
Hey @klickreflex can you privately send me the `SHOPIFY_AUTH_KEY` and `SHOPIFY_AUTH_PASSWORD` so I can run some imports in a test as you? I'm not able to replicate this so far.

**Comment by jackabox** [comment]
Hey @klickreflex, you could try adding some debugging into line [#78 on the ImportSingleProductJob](https://github.com/jackabox/statamic-shopify/blob/1.6.1/src/Jobs/ImportSingleProductJob.php#L78) and see what happens. If you put it inside of the if statement and let me know. 

It seems like the system is failing to find the other Entry that exists, which is odd as the MD files seem basically equivelant.

**Comment by klickreflex** [comment]
Hi @jackabox,

after taking the conversation to emaiI I thought it makes sense to continue here for others to potentially benefit.

In #86 I saw a proposed change to  `SHOPIFY_ASSET_PATH`. I don't have that property in my `.env` at all and I don't find any documentation about it. Is it required or at least recommended to add?

Anything new since I sent you my credentials?

**Comment by jackabox** [comment]
Hey @klickreflex, haven't had the time to spin it up yet. I'm hoping to do so this week. 

Did you manage to insert the debug and see what happened?

Regarding #86 this was for fixing the assets being imported. Are you getting duplicate images or just entries?

**Comment by klickreflex** [comment]
> Did you manage to insert the debug and see what happened?

No, here's my follow-up question on this (sent via email on Nov 8th):

> Regarding the requested debugging output what exactly do I have to do here? Just a var_dump of $entry? I’m not a php dev so I’m not too sure

>  Are you getting duplicate images or just entries?
Only duplicate entries.

**Comment by jackabox** [comment]
Okay, if you run the manual "import products" from the admin/terminal, does this duplicate the entries also?

**Comment by klickreflex** [comment]
Hi Jack,
I added the `var_dump` and ran an import of one of the duplicates, but got no output

![image](https://user-images.githubusercontent.com/435825/143925198-7e0865b7-195e-4b8c-85b8-43d64665fcf7.png)

Instead I've found a new issue:  _Some_ products have variants in Statamic despite not  having any in Shopify (they're all called the same _Default_). I don't even know how to decide which variant to remove manually without maybe breaking anything, let alone preventing this from happening in the future :(
 
![CleanShot 2021-11-29 at 19 50 20@2x](https://user-images.githubusercontent.com/435825/143925614-e50b2db5-ebf9-428e-8136-f676217f7b7c.png)

**Comment by klickreflex** [comment]
When I delete both entries of one that has duplicates and reimport it manually via `php artisan shopify:import:single 7440667738367`  I end up with a duplicate again. ID of one such products is `7440667738367`. Still no var_dump output on the CLI.

**Comment by jackabox** [comment]
Hey @klickreflex thanks for digging deeper into this, I tried replicating it locally but couldn't get any duplication (perhaps because it was all new, or perhaps as no webhook was triggered).

Let me know if the imports have a double effect again after a webhook and if so you could try using a different queue driver, otherwise we'll have to debug some more.

**Comment by jackabox** [comment]
Hey @klickreflex has this normalised now?

**Comment by klickreflex** [comment]
Hi @jackabox 

looks like it didn't happen again after dumping and reimporting everything. Searching for `*.1.md` yields no more results in the products collection.

**Comment by jackabox** [comment]
@klickreflex Odd issue, but glad it's fixed! Going to close this for now, feel free to flag if anything changes.

**Comment by klickreflex** [comment]
Hi @jackabox,

sorry I have to come up with this again. It happened again, this time I have an exact commit with the introduced duplication. 

- Note the `published_at` stamp of both generated items differ for a second. Could it somehow be that Shopify's webhooks fire faster than Statamic can process stuff?
- My client told me she had the product as an unpublished draft first and published later when she found out it didn't show up in Statamic
- A small amount of other new products have been imported without any problems

![CleanShot 2022-01-10 at 19 14 37@2x](https://user-images.githubusercontent.com/435825/148818483-c6bb7974-a869-4dee-b3c4-b17d088d037b.png)

**Comment by klickreflex** [comment]
I've managed to reproduce the issue: duplication happens _always_ when chancing puplication status from draft to published! Could you please have another look @jackabox?

Gonna record a screencast to show the behavior tomorrow.

**Comment by klickreflex** [comment]
Another Update: Now I've had duplication reproducably any time I created products containing 2 or more images, independent of publishing status. Products with one or no images haven't been duplicated thus far. 

Which brings me to the question if the problem could occur because of server resources being exceeded with image generation? There are a couple of failed tasks in Horizon that only seem to run successfully upon a second or so try.

**Comment by jackabox** [comment]
Hey @klickreflex, thanks for this. What queue system are you using?

**Comment by klickreflex** [comment]
Hi @jackabox 

I'm using Redis via Horizon. 

I've send you a Loom video of me doing all sorts of stuff - did this maybe land in your spam (using a new adress and that seem to be flagged often as junk: `klick@hey.com`

**Comment by klickreflex** [comment]
And here are the errors I'm getting for failed jobs in Horizon:


<img width="925" alt="CleanShot 2022-01-14 at 21 23 00@2x" src="https://user-images.githubusercontent.com/435825/149580506-afdb96dd-8d3e-401a-be29-b09624d4bcdb.png">

<img width="932" alt="CleanShot 2022-01-14 at 21 23 08@2x" src="https://user-images.githubusercontent.com/435825/149580491-0e7e07ba-2b73-4e97-9754-3df4f8ad77bf.png">

<img width="923" alt="CleanShot 2022-01-14 at 21 23 15@2x" src="https://user-images.githubusercontent.com/435825/149580484-6b3a9cfc-9b2b-49a7-8797-513e05a6a4e6.png">

With the latest corrupt product imports the entrie Cart stopped working. Can't add products to it and showind only the `ss-cart-loading` message. No JS errors though.

**Comment by klickreflex** [comment]
I've increased server resources, wiped all products and variation from Statamic (again) and triggered a reimport with `php artisan shopify:import:all`:

While this worked and imported products (without duplication, as always when invoked manually), again I'm getting a bunch of failed jobs in Horizon. Here's a full error log:

```
Jackabox\Shopify\Jobs\ImportCollectionsForProductJob
ID
afb8b925-a3f7-4dc9-81eb-6e91417b16da
Queue
default
Attempts
1
Retries
0
Tags
Pushed At
2022-01-16 10:11:26
Failed At
2022-01-16 10:12:02
Exception
Error: Call to a member function selectedQueryColumns() on null in /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Collections/HigherOrderCollectionProxy.php:60
Stack trace:
#0 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php(242): Illuminate\Support\HigherOrderCollectionProxy->Illuminate\Support\{closure}()
#1 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Collections/HigherOrderCollectionProxy.php(61): Illuminate\Support\Collection->each()
#2 /home/forge/stage.crazyceramic.klick.re/vendor/statamic/cms/src/Stache/Query/Builder.php(34): Illuminate\Support\HigherOrderCollectionProxy->__call()
#3 /home/forge/stage.crazyceramic.klick.re/vendor/statamic/cms/src/Query/Builder.php(270): Statamic\Stache\Query\Builder->get()
#4 /home/forge/stage.crazyceramic.klick.re/vendor/statamic/cms/src/Stache/Repositories/EntryRepository.php(40): Statamic\Query\Builder->first()
#5 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(261): Statamic\Stache\Repositories\EntryRepository->find()
#6 /home/forge/stage.crazyceramic.klick.re/vendor/statamic/cms/src/Entries/Entry.php(303): Illuminate\Support\Facades\Facade::__callStatic()
#7 /home/forge/stage.crazyceramic.klick.re/vendor/jackabox/statamic-shopify/src/Jobs/ImportCollectionsForProductJob.php(52): Statamic\Entries\Entry->save()
#8 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Jackabox\Shopify\Jobs\ImportCollectionsForProductJob->handle()
#9 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/Util.php(40): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#10 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#11 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#12 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/Container.php(653): Illuminate\Container\BoundMethod::call()
#13 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(128): Illuminate\Container\Container->call()
#14 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\Bus\Dispatcher->Illuminate\Bus\{closure}()
#15 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#16 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(132): Illuminate\Pipeline\Pipeline->then()
#17 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(120): Illuminate\Bus\Dispatcher->dispatchNow()
#18 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\Queue\CallQueuedHandler->Illuminate\Queue\{closure}()
#19 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#20 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(122): Illuminate\Pipeline\Pipeline->then()
#21 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(70): Illuminate\Queue\CallQueuedHandler->dispatchThroughMiddleware()
#22 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(98): Illuminate\Queue\CallQueuedHandler->call()
#23 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(428): Illuminate\Queue\Jobs\Job->fire()
#24 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(378): Illuminate\Queue\Worker->process()
#25 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(172): Illuminate\Queue\Worker->runJob()
#26 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(117): Illuminate\Queue\Worker->daemon()
#27 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(101): Illuminate\Queue\Console\WorkCommand->runWorker()
#28 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/horizon/src/Console/WorkCommand.php(51): Illuminate\Queue\Console\WorkCommand->handle()
#29 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Laravel\Horizon\Console\WorkCommand->handle()
#30 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/Util.php(40): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#31 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#32 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#33 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Container/Container.php(653): Illuminate\Container\BoundMethod::call()
#34 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Console/Command.php(136): Illuminate\Container\Container->call()
#35 /home/forge/stage.crazyceramic.klick.re/vendor/symfony/console/Command/Command.php(298): Illuminate\Console\Command->execute()
#36 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Console/Command.php(121): Symfony\Component\Console\Command\Command->run()
#37 /home/forge/stage.crazyceramic.klick.re/vendor/symfony/console/Application.php(1005): Illuminate\Console\Command->run()
#38 /home/forge/stage.crazyceramic.klick.re/vendor/symfony/console/Application.php(299): Symfony\Component\Console\Application->doRunCommand()
#39 /home/forge/stage.crazyceramic.klick.re/vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application->doRun()
#40 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Console/Application.php(94): Symfony\Component\Console\Application->run()
#41 /home/forge/stage.crazyceramic.klick.re/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\Console\Application->run()
#42 /home/forge/stage.crazyceramic.klick.re/artisan(37): Illuminate\Foundation\Console\Kernel->handle()
#43 {main}
Data
{
"product": {
"id": "1219bebf-b0cf-4a36-ab37-9d7842c05fb6",
"collection": "products",
"blueprint": null,
"date": null,
"locale": null,
"localizations": null,
"afterSaveCallbacks": [
],
"withEvents": true,
"slug": "butterdose-igel",
"data": {
"items": {
"product_id": 7514730201343,
"published": true,
"published_at": "2022-01-16 09:11:20",
"title": "Butterdose Igel",
"content": "<p data-mce-fragment="1">Sehr süsse , kleine Butterdose für ein halbes Pfund Butter.</p> <p data-mce-fragment="1">Eine Freude für jeden Frühstückstisch!</p>",
"vendor": [
"crazyceramic"
],
"type": null,
"tags": null,
"options": [
],
"featured_image": "shopify/butter_igel_1.jpg",
"gallery": [
"shopify/butter_igel_1.jpg",
"shopify/butter_igel_2.jpg"
]
},
"escapeWhenCastingToString": false
},
"supplements": {
"items": [
],
"escapeWhenCastingToString": false
},
"initialPath": "/home/forge/stage.crazyceramic.klick.re/content/collections/products/butterdose-igel.md",
"published": true,
"selectedQueryColumns": null,
"origin": null
},
"collections": [
{
"id": 287386239231,
"handle": "morgenwonne",
"title": "Morgenwonne",
"updated_at": "2022-01-15T15:17:02+01:00",
"body_html": "",
"published_at": "2021-10-14T15:56:19+02:00",
"sort_order": "best-selling",
"template_suffix": "",
"published_scope": "web",
"admin_graphql_api_id": "gid://shopify/Collection/287386239231"
}
],
"job": null,
"connection": null,
"queue": null,
"chainConnection": null,
"chainQueue": null,
"chainCatchCallbacks": null,
"delay": null,
"afterCommit": null,
"middleware": [
],
"chained": [
]
}
```

**Comment by klickreflex** [comment]
@jackabox 
Could this be a caching issue? I've found this: https://benfurfie.co.uk/articles/how-to-fix-the-call-to-a-member-function-selectedquerycolumns-on-null-error-in-statamic-3

**Comment by klickreflex** [comment]
Yet Aanother observation:

I created one new product in Shopify, is it expected that `ImportCollectionsForProductJob` runs twice? With the same one second delay that the duplicate items show?

![CleanShot 2022-01-16 at 11 54 34@2x](https://user-images.githubusercontent.com/435825/149657047-8ab2f937-84ab-4bad-9b44-4eb032d9ebac.png)

**Comment by klickreflex** [comment]
I was pointed to this core issue that might be the cause of all issues:
https://github.com/statamic/cms/issues/2993

**Comment by klickreflex** [comment]
After removing the Worker from forge I’m having no more duplicate imports caused by the shopify new product Webhook. However, running the CLI command php artisan shopify:import:all doesn’t do anything anymore then:

![image](https://user-images.githubusercontent.com/435825/149659256-26381acc-6810-44be-94cb-b13aa829cfef.png)

**Comment by klickreflex** [comment]
Last observation for today:

- I've setup redis & horizon locally, too to check if I'm getting the same duplication
- When importing a new product manually eg (php artisan shopify:import:product 7515993800959) it's being queued and imported. I saw no duplication locally
- On Staging I ran the exact same single import command end ended up with a duplication

**Comment by j6s** [comment]
@klickreflex and me took a detailed look at the server. We noticed the following things:

* When adding multiple images in shopify, shopify will actually trigger 2 webhooks: 1 for the creation including only the main image and a update webhook containing all other images as well
* These webhooks are dispatched at roughly the same time (usually with at max. 1 second difference)
* Each of those webhooks will dispatch a Job to the queue
* If your queue has multiple workers running then these Jobs will be executed in parallel

Switching to a single queue worker fixes the issue - however this is not ideal since this will throttle all jobs in the system. Ideally, the dispatched jobs should not be allowed to run in parallel. I am not 100% familiar with how Jobs in laravel work, but from my quick research, using the `WithoutOverlapping` middleware on the Jobs could solve this.

**Comment by jackabox** [comment]
Thanks for the deep dive into this. I'll investigate using the details you have provided.

**Comment by jackabox** [comment]
@j6s / @klickreflex Digging into all of this a bit more.

There are a few things that aren't related here, apologies if I miss anything.

### Limiting The Issue

What's the difference between your staging and your local environment? If you're not having the issue locally, but are on staging, this sounds like your configuration is different in some way. Whether this is PHP, dependencies installed, queue processes, etc.

### Caching Issue

```
ErrorException: file_put_contents(/home/forge/stage.crazyceramic.klick
```

This error sounds like it is due to a caching issue, if you clear things from your end this should go away. Don't forget to restart the queue (horizon, etc) whenever you make changes also.

### Regarding Processing Jobs

The jobs are only built to be on one process as they have to queue after each other to ensure that things are working as expected. When you are getting duplicate IDs, do you have multiple processes on the queue? `WithoutOverlapping` won't provide what we want here, but I'm looking into the ability to set a dedicated queue for the jobs that you can then set one process to, this would be the fix for this I believe.

**Comment by jackabox** [comment]
1.7.0 has a queue option in the config that will let you set a new queue with 1 process for now. This should make sure that things resolve 1 by 1.
